src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs

```diff
@@ -169,19 +177,62 @@ internal string CreateCacheKey(CachedVaryBy varyBy)
                             value = "null";
                         }
 
-                        builder.Append(";")
+                        builder.Append(KeyDelimiter)
                             .Append(header)
                             .Append("=")
                             .Append(value);
                     }
                 }
-                // TODO: Parse querystring params
+
+                // Vary by query params
+                if (varyBy?.Params.Count > 0)
+                {
+                    // Append a value denoting the start of the query parameter segment of the cache key
+                    builder.Append(KeyDelimiter)
+                        .Append('Q');
+
+                    if (varyBy.Params.Count == 1 && string.Equals(varyBy.Params[0], "*"))
+                    {
+                        // Vary by all available query params
+                        foreach (var query in _httpContext.Request.Query.OrderBy(q => q.Key))
```

Comment:
Strange this comment got deleted, reposting.
A few issues here. 
1) order by case sensitivity? The QueryCollection operates on case-insensitive keys so maybe we need to order by IgnoreCase and convert all query keys to upper/lower case. Sigh perf...
2) Linq vs Array.Sort() or List.Sort(). Was using Linq to guarantee a stable sort but I think the keys are unique so it doesn't matter. Need to check the impact on performance.
3) Could consider using the raw QueryString but that approach will forgo order-insensitivity of the params. I don't think we should do this unless perf is a big problem here.